src/lightning_app/components/multi_node/pytorch_spawn.py

```diff
@@ -31,10 +34,31 @@ def __call__(
     ):
         import torch
 
-        nprocs = torch.cuda.device_count() if torch.cuda.is_available() else 1
-        torch.multiprocessing.spawn(
-            self.run, args=(self.work_run, main_address, main_port, num_nodes, node_rank, nprocs), nprocs=nprocs
-        )
+        with self.enable_spawn():
+            nprocs = torch.cuda.device_count() if torch.cuda.is_available() else 1
+            queue = self.delta_queue if isinstance(self.delta_queue, MultiProcessQueue) else self.delta_queue.to_dict()
+            torch.multiprocessing.spawn(
+                self.dispatch_run,
+                args=(self.__class__, self.work, queue, main_address, main_port, num_nodes, node_rank, nprocs),
+                nprocs=nprocs,
+            )
+
+    @staticmethod
+    def dispatch_run(local_rank, cls, work, delta_queue, *args, **kwargs):
+        if local_rank == 0:
+            if isinstance(delta_queue, dict):
+                delta_queue = cls.process_queue(delta_queue)
+                work._request_queue = cls.process_queue(work._request_queue)
+                work._response_queue = cls.process_queue(work._response_queue)
+
+            state_observer = WorkStateObserver(work, delta_queue=delta_queue)
```

Comment:
Just an observation: here we are destroying WorkStateObserver when run returns, and eventually instantiate a new one on a subsequent call to run.
However, that run method might be starting up a thread that modifies the state and return immediately, and we would lose the ability to detect state changes resulting from that.
Wouldn’t it be better to initialize the observer during the first call to run, and then keep it running until the work is stopped?

Reply:
Yes, 100 %. I faced some concurrency issues in the past, but we can definitely re-explore it. This would definitely be the right way.